src/station_cmd.cpp

```diff
@@ -2535,14 +2535,15 @@ CommandCost CmdBuildDock(DoCommandFlag flags, TileIndex tile, StationID station_
 
 	if (flags & DC_EXEC) {
 		st->ship_station.Add(tile);
-		st->ship_station.Add(tile + TileOffsByDiagDir(direction));
+		TileIndex flat_tile = tile + TileOffsByDiagDir(direction);
+		st->ship_station.Add(flat_tile);
 		st->AddFacility(FACIL_DOCK, tile);
 
 		st->rect.BeforeAddRect(dock_area.tile, dock_area.w, dock_area.h, StationRect::ADD_TRY);
 
 		/* If the water part of the dock is on a canal, update infrastructure counts.
-		 * This is needed as we've unconditionally cleared that tile before. */
-		if (wc == WATER_CLASS_CANAL) {
+		 * This is needed as we've cleared that tile before. */
+		if (wc == WATER_CLASS_CANAL && !(HasTileWaterClass(flat_tile) && GetWaterClass(flat_tile) == WATER_CLASS_CANAL && IsTileOwner(flat_tile, _current_company))) {
```

Comment:
Oh, how fun that `ClearTile` for objects does not actually clear the tile but removes the object. That might be messing more things up; maybe also locks?
Maybe at add a note in the comment in these places that for `MP_OBJECT` clearing could yield a `MP_WATER` tile instead of `MP_CLEAR`? I got very confused by this, as it did clear so the tile should have been `MP_CLEAR` and as such the `HasTileWaterClass` would already/still be returning false.